File: sdks/kotlin/src/test/kotlin/io/github/mikemackintosh/crowdcontrol/EvaluatorTest.kt
package io.github.mikemackintosh.crowdcontrol

import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFalse
import kotlin.test.assertTrue

class EvaluatorTest {

    @Test
    fun forbidFiresWhenConditionMatches() {
        val eng = fromSource(
            listOf(
                """
                forbid "no-public" {
                    resource.acl == "public-read"
                    message "bucket is public"
                }
                """.trimIndent()
            )
        )
        val results = eng.evaluate(mapOf("resource" to mapOf("acl" to "public-read")))
        assertEquals(1, results.size)
        assertEquals("no-public", results[0].rule)
        assertFalse(results[0].passed)
        assertTrue(results[0].message.contains("bucket is public"))
    }

    @Test
    fun forbidDoesNotFireWhenMismatch() {
        val eng = fromSource(
            listOf(
                """
                forbid "no-public" {
                    resource.acl == "public-read"
                    message "bucket is public"
                }
                """.trimIndent()
            )
        )
        val results = eng.evaluate(mapOf("resource" to mapOf("acl" to "private")))
        assertTrue(results[0].passed)
    }

    @Test
    fun permitFires() {
        val eng = fromSource(
            listOf(
                """
                permit "admin-ok" {
                    user.role == "admin"
                    message "approved"
                }
                """.trimIndent()
            )
        )
        val results = eng.evaluate(mapOf("user" to mapOf("role" to "admin")))
        assertTrue(results[0].passed)
        assertEquals("approved", results[0].message)
    }

    @Test
    fun unlessSavesRule() {
        val eng = fromSource(
            listOf(
                """
                forbid "rule" {
                    user.role == "intern"
                    unless user.approved == true
                    message "intern"
                }
                """.trimIndent()
            )
        )
        val results = eng.evaluate(mapOf("user" to mapOf("role" to "intern", "approved" to true)))
        assertTrue(results[0].passed)
    }

    @Test
    fun numericComparisonsOnlyWorkOnNumbers() {
        val eng = fromSource(listOf("""forbid "r" { user.age < 18 }"""))
        val r1 = eng.evaluate(mapOf("user" to mapOf("age" to 16)))
        assertFalse(r1[0].passed)
        val r2 = eng.evaluate(mapOf("user" to mapOf("age" to 21)))
        assertTrue(r2[0].passed)
        // String values should not coerce: "16" < 18 is false.
        val r3 = eng.evaluate(mapOf("user" to mapOf("age" to "16")))
        assertTrue(r3[0].passed)
    }

    @Test
    fun containsOnListAndString() {
        val eng = fromSource(
            listOf(
                """
                forbid "r" {
                    labels contains "danger"
                }
                """.trimIndent()
            )
        )
        val hit = eng.evaluate(mapOf("labels" to listOf("ok", "danger", "more")))
        assertFalse(hit[0].passed)
        val miss = eng.evaluate(mapOf("labels" to listOf("ok")))
        assertTrue(miss[0].passed)
    }

    @Test
    fun anyQuantifier() {
        val eng = fromSource(
            listOf(
                """
                forbid "r" {
                    any resources matches "prod-*"
                }
                """.trimIndent()
            )
        )
        val hit = eng.evaluate(mapOf("resources" to listOf("dev-db", "prod-api")))
        assertFalse(hit[0].passed)
        val miss = eng.evaluate(mapOf("resources" to listOf("dev-db")))
        assertTrue(miss[0].passed)
    }

    @Test
    fun allQuantifierEmptyListIsTrue() {
        val eng = fromSource(
            listOf(
                """
                forbid "r" {
                    all items == "bad"
                }
                """.trimIndent()
            )
        )
        val results = eng.evaluate(mapOf("items" to emptyList<String>()))
        assertFalse(results[0].passed) // all vacuously true -> rule fires
    }

    @Test
    fun countAggregate() {
        val eng = fromSource(listOf("""forbid "r" { count(items) > 2 }"""))
        val hit = eng.evaluate(mapOf("items" to listOf(1, 2, 3, 4)))
        assertFalse(hit[0].passed)
        val miss = eng.evaluate(mapOf("items" to listOf(1)))
        assertTrue(miss[0].passed)
    }

    @Test
    fun arithmeticExpr() {
        val eng = fromSource(listOf("""forbid "r" { used + requested > capacity }"""))
        val hit = eng.evaluate(mapOf("used" to 80, "requested" to 30, "capacity" to 100))
        assertFalse(hit[0].passed)
        val miss = eng.evaluate(mapOf("used" to 10, "requested" to 10, "capacity" to 100))
        assertTrue(miss[0].passed)
    }

    @Test
    fun messageInterpolation() {
        val eng = fromSource(
            listOf(
                """
                forbid "r" {
                    user.role == "intern"
                    message "user {user.name} is an intern"
                }
                """.trimIndent()
            )
        )
        val results = eng.evaluate(mapOf("user" to mapOf("name" to "alex", "role" to "intern")))
        assertEquals("user alex is an intern", results[0].message)
    }

    @Test
    fun defaultDenyAppendsImplicitResult() {
        val eng = fromSource(
            listOf(
                """
                permit "specific" {
                    user.role == "admin"
                    message "approved"
                }
                """.trimIndent()
            ),
            defaultEffect = DefaultEffect.DENY,
        )
        val results = eng.evaluate(mapOf("user" to mapOf("role" to "developer")))
        assertEquals(2, results.size)
        assertEquals("(default-deny)", results[1].rule)
        assertFalse(results[1].passed)
    }

    @Test
    fun negation() {
        val eng = fromSource(listOf("""forbid "r" { not user.approved == true }"""))
        val denied = eng.evaluate(mapOf("user" to mapOf("approved" to false)))
        assertFalse(denied[0].passed)
        val ok = eng.evaluate(mapOf("user" to mapOf("approved" to true)))
        assertTrue(ok[0].passed)
    }

    @Test
    fun orGroup() {
        val eng = fromSource(
            listOf(
                """
                forbid "r" {
                    user.role == "intern" or user.role == "contractor"
                }
                """.trimIndent()
            )
        )
        val a = eng.evaluate(mapOf("user" to mapOf("role" to "intern")))
        assertFalse(a[0].passed)
        val b = eng.evaluate(mapOf("user" to mapOf("role" to "contractor")))
        assertFalse(b[0].passed)
        val c = eng.evaluate(mapOf("user" to mapOf("role" to "admin")))
        assertTrue(c[0].passed)
    }

    @Test
    fun lowerTransform() {
        val eng = fromSource(listOf("""forbid "r" { lower(user.name) == "root" }"""))
        val hit = eng.evaluate(mapOf("user" to mapOf("name" to "ROOT")))
        assertFalse(hit[0].passed)
    }

    @Test
    fun resolveFieldReturnsNullOnMissing() {
        val doc = mapOf<String, Any?>("a" to mapOf("b" to 1))
        assertEquals(1, resolveField("a.b", doc))
        assertEquals(null, resolveField("a.c", doc))
        assertEquals(null, resolveField("x.y", doc))
    }

    @Test
    fun fmtVMatchesGoSemantics() {
        assertEquals("<nil>", fmtV(null))
        assertEquals("true", fmtV(true))
        assertEquals("false", fmtV(false))
        assertEquals("42", fmtV(42))
        assertEquals("42", fmtV(42.0))
        assertEquals("3.14", fmtV(3.14))
        assertEquals("hello", fmtV("hello"))
    }
}

File: sdks/typescript/tests/evaluator.test.ts
import { test } from "node:test";
import assert from "node:assert/strict";

import {
  Evaluator,
  fmtV,
  formatResults,
  fromSource,
  interpolateMessage,
  resolveField,
} from "../src/evaluator.js";
import { DEFAULT_DENY } from "../src/types.js";

function evalPolicy(src: string, doc: Record<string, unknown>) {
  return fromSource([src]).evaluate(doc);
}

test("eval: basic forbid fires when condition true", () => {
  const results = evalPolicy(
    'forbid "x" { user.role == "admin" message "blocked" }',
    { user: { role: "admin" } },
  );
  assert.equal(results.length, 1);
  assert.equal(results[0]!.passed, false);
  assert.equal(results[0]!.message, "blocked");
});

test("eval: forbid does not fire when condition false", () => {
  const results = evalPolicy(
    'forbid "x" { user.role == "admin" message "blocked" }',
    { user: { role: "dev" } },
  );
  assert.equal(results[0]!.passed, true);
  assert.equal(results[0]!.message, "");
});

test("eval: permit result has passed true and message set when fired", () => {
  const results = evalPolicy(
    'permit "ok" { user.role == "admin" message "allowed" }',
    { user: { role: "admin" } },
  );
  assert.equal(results[0]!.kind, "permit");
  assert.equal(results[0]!.passed, true);
  assert.equal(results[0]!.message, "allowed");
});

test("eval: unless saves rule", () => {
  const results = evalPolicy(
    'forbid "r" { a == 1 unless b == 2 message "fail" }',
    { a: 1, b: 2 },
  );
  assert.equal(results[0]!.passed, true);
});

test("eval: multiple conditions are ANDed", () => {
  const results = evalPolicy(
    'forbid "r" { a == 1 b == 2 message "x" }',
    { a: 1, b: 3 },
  );
  assert.equal(results[0]!.passed, true); // b mismatch → rule does not fire
});

test("eval: any quantifier true when an element matches", () => {
  const results = evalPolicy(
    'forbid "r" { any labels == "prod" message "prod" }',
    { labels: ["dev", "prod", "test"] },
  );
  assert.equal(results[0]!.passed, false);
});

test("eval: all quantifier on empty list is vacuously true", () => {
  const results = evalPolicy(
    'forbid "r" { all labels == "x" message "fail" }',
    { labels: [] },
  );
  assert.equal(results[0]!.passed, false);
});

test("eval: count aggregate", () => {
  const results = evalPolicy(
    'forbid "r" { count(deletes) > 3 message "too many" }',
    { deletes: [1, 2, 3, 4, 5] },
  );
  assert.equal(results[0]!.passed, false);
});

test("eval: matches glob", () => {
  const results = evalPolicy(
    'forbid "r" { name matches "prod-*" message "prod" }',
    { name: "prod-db" },
  );
  assert.equal(results[0]!.passed, false);
});

test("eval: matches_regex", () => {
  const results = evalPolicy(
    'forbid "r" { name matches_regex "^[a-z]+$" message "lower" }',
    { name: "abc" },
  );
  assert.equal(results[0]!.passed, false);
});

test("eval: in list", () => {
  const results = evalPolicy(
    'forbid "r" { role in ["admin", "root"] message "priv" }',
    { role: "admin" },
  );
  assert.equal(results[0]!.passed, false);
});

test("eval: contains on list and string", () => {
  const r1 = evalPolicy('forbid "a" { tags contains "x" message "m" }', {
    tags: ["a", "x", "b"],
  });
  assert.equal(r1[0]!.passed, false);
  const r2 = evalPolicy('forbid "a" { name contains "abc" message "m" }', {
    name: "xabcx",
  });
  assert.equal(r2[0]!.passed, false);
});

test("eval: intersects and is_subset", () => {
  const r1 = evalPolicy('forbid "r" { tags intersects ["x", "y"] message "m" }', {
    tags: ["y"],
  });
  assert.equal(r1[0]!.passed, false);
  const r2 = evalPolicy('forbid "r" { tags is_subset ["a", "b", "c"] message "m" }', {
    tags: ["a", "b"],
  });
  assert.equal(r2[0]!.passed, false);
});

test("eval: message interpolation", () => {
  const msg = interpolateMessage("hello {user.name}", { user: { name: "ana" } });
  assert.equal(msg, "hello ana");
});

test("eval: message interpolation with count()", () => {
  const msg = interpolateMessage("{count(items)} items", { items: [1, 2, 3] });
  assert.equal(msg, "3 items");
});

test("eval: default-deny appended when no rules match", () => {
  const eng = fromSource(['permit "p" { user.role == "admin" message "ok" }'], {
    defaultEffect: DEFAULT_DENY,
  });
  const results = eng.evaluate({ user: { role: "dev" } });
  assert.equal(results.length, 2);
  assert.equal(results[1]!.rule, "(default-deny)");
  assert.equal(results[1]!.passed, false);
});

test("eval: default-deny suppressed when permit fires", () => {
  const eng = fromSource(['permit "p" { user.role == "admin" message "ok" }'], {
    defaultEffect: DEFAULT_DENY,
  });
  const results = eng.evaluate({ user: { role: "admin" } });
  assert.equal(results.length, 1);
});

test("eval: resolveField navigates nested maps", () => {
  assert.equal(resolveField("a.b.c", { a: { b: { c: 42 } } }), 42);
  assert.equal(resolveField("a.x", { a: {} }), null);
  assert.equal(resolveField("a.b", { a: "not a map" }), null);
});

test("eval: fmtV matches Go %v semantics", () => {
  assert.equal(fmtV(true), "true");
  assert.equal(fmtV(false), "false");
  assert.equal(fmtV(null), "<nil>");
  assert.equal(fmtV(undefined), "<nil>");
  assert.equal(fmtV(42), "42");
  assert.equal(fmtV(3.14), "3.14");
  assert.equal(fmtV("hi"), "hi");
});

test("eval: format_results denies and passes", () => {
  const results = evalPolicy(
    'forbid "x" { a == 1 message "boom" }',
    { a: 1 },
  );
  const { text, allPassed } = formatResults(results);
  assert.equal(allPassed, false);
  assert.ok(text.includes("DENY: boom (x)"));
});

test("eval: arithmetic expression condition", () => {
  const results = evalPolicy(
    'forbid "r" { plan.adds + plan.deletes > 10 message "m" }',
    { plan: { adds: 6, deletes: 5 } },
  );
  assert.equal(results[0]!.passed, false);
});

test("eval: transform lower()", () => {
  const results = evalPolicy(
    'forbid "r" { lower(user.name) == "alex" message "m" }',
    { user: { name: "ALEX" } },
  );
  assert.equal(results[0]!.passed, false);
});

test("eval: Evaluator class methods", () => {
  const eng = new Evaluator();
  assert.deepEqual(eng.policies(), []);
  assert.deepEqual(eng.evaluate({}), []);
});
